# Focusing on the drugs in the infer_drug.tsv file

### Get the mapping of SMILES and DrugBankID

In [2]:
# read drugbank_smiles.txt
import numpy as np
import pandas as pd
import csv
from io import StringIO 
import torch as th

file_path = "../drugbank_info/drugbank_smiles.txt"

smiles_mapping = dict()

# Read the text file line by line and populate the dictionary
with open(file_path, 'r') as file:
    for line in file:
        # Split each line into DrugBankID and SMILES using tab as the delimiter
        drugbank_id, smiles = line.strip().split('\t')
        
        # Add the entry to the dictionary
        smiles_mapping[drugbank_id] = smiles

# Display the dictionaries
# print(smiles_mapping)

In [3]:
# Get the number of available drugs in the drug bank
len(smiles_mapping)

8807

There are 8807 compounds in the `drugbank_smiles.txt` file.

NOTE: Not all the drugs here are in the `infer_drug.tsv` file. E.g `DB14708, DB00114, DB00117, DB00119, DB00120, DB00121, ... `
The `infer_drug.tsv` only has drugs with a molecular weight >= 250.

## Get the embeddings of each SMILES

In [4]:
# Load entity file
entity_idmap_file = '../data/embed/entities.tsv'

# Get all the drugs on a list 
# drug_list = list(df['DrugBankID'])

# Load entity file
drug_list = []
with open("./infer_drug.tsv", newline='', encoding='utf-8') as csvfile:
    reader = csv.DictReader(csvfile, delimiter='\t', fieldnames=['drug','ids'])
    for row_val in reader:
        drug_list.append(row_val['drug'])

print("There are", len(drug_list), "drugs in the infer_drug file")

There are 8104 drugs in the infer_drug file


In [5]:
# Get drug name to entity ID mappings
entity_map = {}

with open(entity_idmap_file, newline='', encoding='utf-8') as csvfile:
    reader = csv.DictReader(csvfile, delimiter='\t', fieldnames=['name','id'])
    for row_val in reader:
        entity_map[row_val['name']] = int(row_val['id'])
 
# handle the ID mapping
drug_ids = []
drug_entity = {}  # mapping for drugs to entity for drugs in the DRKG only 
# disease_ids = []
for drug in drug_list:
    drug_entity.update({drug:entity_map[drug]})
    drug_ids.append(entity_map[drug])

In [6]:
# preview the dictionary
print({k: drug_entity[k] for k in list(drug_entity)[:5]})

# get the length of the dictionary
print(len(drug_entity))

{'Compound::DB00605': 9475, 'Compound::DB00983': 11010, 'Compound::DB01240': 7486, 'Compound::DB11755': 16376, 'Compound::DB12184': 4505}
8104


There are 8104 drugs in the `drug_entity` mapping. All the drugs have been mapped.

In [7]:
# Create a dataframe of the drug_entity dictionary
df = pd.DataFrame(list(drug_entity.items()), 
                  columns=['drug_bank_id', 'entity_id'])
df['drug_bank_id'] = df['drug_bank_id'].str.replace('Compound::', '')
df.head()

drug_bank_id  entity_id
0      DB00605       9475
1      DB00983      11010
2      DB01240       7486
3      DB11755      16376
4      DB12184       4505

In [8]:
# Adding the SMILES column based on the smiles_mapping dictionary
df['SMILES'] = df['drug_bank_id'].map(smiles_mapping)

# Handle NaN values and convert 'SMILES' column to integers and 
# df['SMILES'] = df['SMILES'].fillna(-1).astype(int)

df.head()

drug_bank_id  entity_id                                             SMILES
0      DB00605       9475  CC1=C(CC(O)=O)C2=CC(F)=CC=C2\C1=C/C1=CC=C(C=C1...
1      DB00983      11010     COC1=CC=C(CC(C)NCC(O)C2=CC(NC=O)=C(O)C=C2)C=C1
2      DB01240       7486  [H][C@]12C[C@@H](O)[C@H](\C=C\[C@@H](O)CCCCC)[...
3      DB11755      16376   CCCC1=CC(O)=C2[C@@H]3C=C(C)CC[C@H]3C(C)(C)OC2=C1
4      DB12184       4505    CC1(C)CC(=O)N(CCCCN2CCN(CC2)C2=NC=CC=N2)C(=O)C1

In [9]:
# Check for missing values (NaN) in the 'SMILES' column
df[df['SMILES'].isna()]

drug_bank_id  entity_id SMILES
14        DB12274      15567    NaN
15        DB13375      15614    NaN
27        DB05325      15694    NaN
38        DB00026      10157    NaN
51        DB01277      10394    NaN
...           ...        ...    ...
8077      DB14793       5019    NaN
8078      DB09108      24669    NaN
8082      DB12009      14702    NaN
8084      DB13900      25019    NaN
8093      DB11836      11419    NaN

[1583 rows x 3 columns]

There are 1583 missing values of SMILES. We will filter them out.

In [10]:
df = df[df['SMILES'].notna()]
print(df.shape)
df.head()

(6521, 3)


drug_bank_id  entity_id                                             SMILES
0      DB00605       9475  CC1=C(CC(O)=O)C2=CC(F)=CC=C2\C1=C/C1=CC=C(C=C1...
1      DB00983      11010     COC1=CC=C(CC(C)NCC(O)C2=CC(NC=O)=C(O)C=C2)C=C1
2      DB01240       7486  [H][C@]12C[C@@H](O)[C@H](\C=C\[C@@H](O)CCCCC)[...
3      DB11755      16376   CCCC1=CC(O)=C2[C@@H]3C=C(C)CC[C@H]3C(C)(C)OC2=C1
4      DB12184       4505    CC1(C)CC(=O)N(CCCCN2CCN(CC2)C2=NC=CC=N2)C(=O)C1

In [11]:
df.tail()

drug_bank_id  entity_id  \
8099      DB03445      22548   
8100      DB11200      24747   
8101      DB01897      21144   
8102      DB08404      24473   
8103      DB05455      23655   

                                                 SMILES  
8099  [H]\C(N[C@@]([H])(C(O)=O)[C@](C)(CN1C=CN=N1)S(...  
8100                O.O.[Al+3].[Cl-].[Zr+4].NCC([O-])=O  
8101  O=C1NNC(=O)C2=CC(=CC=C12)[N+]1=NC(\C=C\C2=CC=C...  
8102  [H][C@](O)(C(=O)NCCC(=O)NCCSC(=O)CCCCC)C(C)(C)...  
8103  CCOC1=CC2=C(C=C1)N(C(=O)C23CCC(CC3)OCCN4CCOCC4...

There are `6521` drugs from the `infer_drugs` files that have a SMILES value from the `drugbank_smiles.txt` file.

In [12]:
drug_ids[:5]

[9475, 11010, 7486, 16376, 4505]

The `drug_ids` list values are in the same order as the `entity_id` column in our dataframe.
It will be easy to concatenate the embedding of each drug to the dataframe based on the `entity_id.`

In [13]:
len(drug_ids)

8104

The lenght of `drug_ids` is still `8104`. We need to remove the drug_ids that are not in our dataframe.

In [14]:
# Get only the drug_ids that are in the dataframe
drug_ids = [k for k in drug_ids if k in set(df['entity_id'])]

# check
print(len(drug_ids))

6521


In [15]:
drug_ids[:10]

[9475, 11010, 7486, 16376, 4505, 7417, 6457, 7161, 2415, 4943]

Now the lenght of drug_ids match the lenght of the dataframe.

In [16]:
# Load embedding
entity_emb = np.load('../data/embed/DRKG_TransE_l2_entity.npy')

drug_ids = th.tensor(drug_ids).long()
drug_emb = th.tensor(entity_emb[drug_ids])

In [17]:
# view the entity embeddings
entity_emb

array([[-0.5888741 , -0.0872229 , -0.45159575, ...,  0.6187399 ,
        -0.51274306,  0.43644685],
       [ 0.03043327, -0.9927482 ,  0.31472597, ..., -0.03531377,
        -0.7016256 , -0.07200253],
       [-0.15260178, -0.31713626, -0.3052066 , ..., -0.5843045 ,
        -0.5971722 ,  0.4716604 ],
       ...,
       [ 0.58301383,  0.69486225,  0.5986209 , ...,  0.58480304,
         0.5156303 ,  0.27735808],
       [ 0.7441235 , -0.58081806,  0.3950408 , ...,  0.56884754,
        -0.6283229 ,  0.5594293 ],
       [-0.6101573 , -0.4708205 , -0.4374642 , ...,  0.42074913,
        -0.56538516,  0.621319  ]], dtype=float32)

In [18]:
# view only the drug embeddings, which is a subset of the entity embeddings
print(len(drug_emb))
drug_emb

6521


tensor([[-0.2715, -0.5940, -0.3701,  ..., -0.5073,  0.1592, -0.6702],
        [-0.4294, -0.3552, -0.4526,  ...,  0.6304,  0.4417, -0.4394],
        [-0.6724, -0.2223, -0.5301,  ...,  0.3787,  0.3645, -0.3003],
        ...,
        [ 0.1658,  0.6800, -0.5095,  ..., -0.6383, -0.5285, -0.6048],
        [ 0.5265,  0.6320, -0.5831,  ...,  0.5215, -0.5397,  0.5052],
        [-0.5288, -0.4380, -0.6211,  ..., -0.5878, -0.5739, -0.5481]])

In [19]:
# Convert the tensor to a Pandas DataFrame
emb_df = pd.DataFrame(drug_emb.numpy(), columns=[f'embedding_{i}' for i in range(400)])

# Concatenate the new DataFrame with the existing DataFrame
result_df = pd.concat([df.reset_index(drop=True), emb_df], axis=1)

# Display the updated DataFrame
result_df.head(10)

drug_bank_id  entity_id                                             SMILES  \
0      DB00605       9475  CC1=C(CC(O)=O)C2=CC(F)=CC=C2\C1=C/C1=CC=C(C=C1...   
1      DB00983      11010     COC1=CC=C(CC(C)NCC(O)C2=CC(NC=O)=C(O)C=C2)C=C1   
2      DB01240       7486  [H][C@]12C[C@@H](O)[C@H](\C=C\[C@@H](O)CCCCC)[...   
3      DB11755      16376   CCCC1=CC(O)=C2[C@@H]3C=C(C)CC[C@H]3C(C)(C)OC2=C1   
4      DB12184       4505    CC1(C)CC(=O)N(CCCCN2CCN(CC2)C2=NC=CC=N2)C(=O)C1   
5      DB00404       7417     CC1=NN=C2CN=C(C3=CC=CC=C3)C3=C(C=CC(Cl)=C3)N12   
6      DB01223       6457  NCCN.CN1C2=C(NC=N2)C(=O)N(C)C1=O.CN1C2=C(NC=N2...   
7      DB00572       7161  CN1[C@H]2CC[C@@H]1C[C@@H](C2)OC(=O)C(CO)C1=CC=...   
8      DB00669       2415             CNS(=O)(=O)CC1=CC2=C(NC=C2CCN(C)C)C=C1   
9      DB00494       4943  CCN(CC)C(=O)C(=C\C1=CC(=C(O)C(O)=C1)[N+]([O-])...   

   embedding_0  embedding_1  embedding_2  embedding_3  embedding_4  \
0    -0.271492    -0.593986    -0.370118    -0.132015     0.647420   
1    -0.429383    -0.355152    -0.452635    -0.493196    -0.350862   
2    -0.672433    -0.222313    -0.530109    -0.546897     0.308746   
3    -0.369791     0.667106    -0.406856    -0.587619     0.577712   
4    -0.151366     0.479758    -0.379089    -0.545514     0.743205   
5    -0.589351    -0.580602    -0.745123    -0.528509     0.593207   
6    -0.674336    -0.266500    -0.512340    -0.115845    -0.309642   
7    -0.297685    -0.514792    -0.603773    -0.658746     0.420709   
8    -0.453169    -0.235891    -0.643428    -0.610339     0.023665   
9    -0.535264    -0.502035    -0.528655    -0.100440    -0.082784   

   embedding_5  embedding_6  ...  embedding_390  embedding_391  embedding_392  \
0     0.267383     0.221824  ...      -0.110515       0.747298      -0.595123   
1     0.713262     0.443782  ...      -0.542713       0.260568      -0.455372   
2    -0.337016     0.149878  ...       0.565954      -0.691679      -0.595132   
3     0.643965    -0.662907  ...       0.459833      -0.737107      -0.616846   
4     0.751715    -0.588066  ...       0.427640      -0.665244      -0.284450   
5     0.497097     0.048647  ...      -0.678077      -0.345387      -0.128594   
6     0.626041     0.451178  ...      -0.554611       0.078451      -0.180794   
7     0.737039     0.000725  ...      -0.306882       0.153852      -0.428181   
8    -0.006115    -0.559450  ...       0.119811       0.474479       0.602419   
9     0.540049     0.041261  ...      -0.705202       0.230018      -0.369495   

   embedding_393  embedding_394  embedding_395  embedding_396  embedding_397  \
0      -0.354849       0.426570      -0.651937      -0.535123      -0.507326   
1      -0.012449      -0.440367      -0.572628      -0.547223       0.630401   
2      -0.309585       0.082130      -0.466184      -0.515885       0.378655   
3      -0.169433      -0.602083      -0.120658      -0.194525      -0.377465   
4       0.689696      -0.577339      -0.528556       0.713676      -0.300458   
5       0.402087       0.085832      -0.420954      -0.311380      -0.376174   
6      -0.402600      -0.749248       0.438113      -0.795475      -0.051928   
7      -0.233794      -0.307042      -0.313246      -0.309900       0.096056   
8       0.255796      -0.105829      -0.557198       0.094172       0.142557   
9       0.308531       0.017748      -0.620017       0.475681       0.461897   

   embedding_398  embedding_399  
0       0.159212      -0.670219  
1       0.441735      -0.439394  
2       0.364504      -0.300323  
3       0.231654      -0.369904  
4       0.254337       0.029720  
5       0.388722      -0.459413  
6       0.437143      -0.286584  
7       0.473331      -0.490021  
8       0.284198      -0.518554  
9       0.462243      -0.646051  

[10 rows x 403 columns]

In [20]:
emb_df.shape

(6521, 400)

In [21]:
# Now remove the entity_id column
result_df.drop('entity_id', inplace=True, axis=1)
result_df.head()

drug_bank_id                                             SMILES  \
0      DB00605  CC1=C(CC(O)=O)C2=CC(F)=CC=C2\C1=C/C1=CC=C(C=C1...   
1      DB00983     COC1=CC=C(CC(C)NCC(O)C2=CC(NC=O)=C(O)C=C2)C=C1   
2      DB01240  [H][C@]12C[C@@H](O)[C@H](\C=C\[C@@H](O)CCCCC)[...   
3      DB11755   CCCC1=CC(O)=C2[C@@H]3C=C(C)CC[C@H]3C(C)(C)OC2=C1   
4      DB12184    CC1(C)CC(=O)N(CCCCN2CCN(CC2)C2=NC=CC=N2)C(=O)C1   

   embedding_0  embedding_1  embedding_2  embedding_3  embedding_4  \
0    -0.271492    -0.593986    -0.370118    -0.132015     0.647420   
1    -0.429383    -0.355152    -0.452635    -0.493196    -0.350862   
2    -0.672433    -0.222313    -0.530109    -0.546897     0.308746   
3    -0.369791     0.667106    -0.406856    -0.587619     0.577712   
4    -0.151366     0.479758    -0.379089    -0.545514     0.743205   

   embedding_5  embedding_6  embedding_7  ...  embedding_390  embedding_391  \
0     0.267383     0.221824    -0.669338  ...      -0.110515       0.747298   
1     0.713262     0.443782    -0.526725  ...      -0.542713       0.260568   
2    -0.337016     0.149878    -0.787017  ...       0.565954      -0.691679   
3     0.643965    -0.662907    -0.538321  ...       0.459833      -0.737107   
4     0.751715    -0.588066    -0.573652  ...       0.427640      -0.665244   

   embedding_392  embedding_393  embedding_394  embedding_395  embedding_396  \
0      -0.595123      -0.354849       0.426570      -0.651937      -0.535123   
1      -0.455372      -0.012449      -0.440367      -0.572628      -0.547223   
2      -0.595132      -0.309585       0.082130      -0.466184      -0.515885   
3      -0.616846      -0.169433      -0.602083      -0.120658      -0.194525   
4      -0.284450       0.689696      -0.577339      -0.528556       0.713676   

   embedding_397  embedding_398  embedding_399  
0      -0.507326       0.159212      -0.670219  
1       0.630401       0.441735      -0.439394  
2       0.378655       0.364504      -0.300323  
3      -0.377465       0.231654      -0.369904  
4      -0.300458       0.254337       0.029720  

[5 rows x 402 columns]

In [22]:
result_df.tail(50)

drug_bank_id                                             SMILES  \
6471      DB03064                        CCCCCCCCCCC1=C(O)C(=O)NC1=O   
6472      DB03495  [H][C@]1(C)O[C@]([H])(O[C@]2([H])[C@@]([H])(CO...   
6473      DB02449  [H][C@@](CC1=CNC2=CC=CC=C12)(NS(=O)(=O)C1=CC=C...   
6474      DB04000                N[C@@H](CN1C=C(Br)C(=O)NC1=O)C(O)=O   
6475      DB06988  [H][C@](N)(CC1=CC(\N=C\CC2=CC=CC=C2)=C(O)C=C1O...   
6476      DB06873  OCCOCN1C=C(CC2=CC=CC(OCC3=CC=CC=C3)=C2)C(=O)NC1=O   
6477      DB02790  [H]N1C(=O)C=CN([C@@H]2O[C@H](COP(O)(=O)OP(O)(=...   
6478      DB03558  OC(=O)CCNC(=O)[C@H]1CCCN1S(=O)(=O)C1=CC=C(C=C1...   
6479      DB02133  [Mg++].CCC1=C(C)C2=N\C\1=C/C1=C(C)C3=C([N-]1)\...   
6480      DB04662  [H][C@@](CC)(CO)NC1=NC2=C(N=CN2C(C)C)C(NCC2=CC...   
6481      DB08216       CC(C)(C)C1=CC(=CC=C1O)\N=N\C1=CC=CC=C1C(O)=O   
6482      DB03213  NC(=N)C1=CC2=C(C=C1)N=C(N2)C(=O)C1=NC2=C(N1)C=...   
6483      DB04277       OCC(=O)[C@@H](O)[C@@H](O)[C@H](O)COP(O)(O)=O   
6484      DB08328  [H][C@](O)(C(=O)NCCC(=O)NCCO)C(C)(C)COC(=O)C(C...   
6485      DB02505  CCCCCC1=CC=C(C=C1)C(=O)N(CC1=CC=C(C=C1)C1=CC=C...   
6486      DB07827  CN1C(=O)N(CC2=CC=C(C=C2)C(O)=O)C(=O)C2=C1C=CC(...   
6487      DB04651  NC1=C2N=CN([C@@H]3O[C@H](CO[P@@](O)(=O)OCCCCC[...   
6488      DB04672  [H][C@]1(COP(O)(O)=O)O[C@]([H])(C=C1)N1C=C(C)C...   
6489      DB07219  [H][C@](C)(CC)[C@]([H])(NC(=O)[C@@]1([H])O[C@]...   
6490      DB08719          CC1=NOC(CCCCCOC2=CC=C(C=C2Cl)C2=NCCO2)=C1   
6491      DB11380         CC(C)OC(=O)NC1=CC2=C(NC(=N2)C2=CSC=N2)C=C1   
6492      DB07099                 NCC(=O)NC1=CC=C(OCC2=CC=CC=C2)C=C1   
6493      DB04733  O[C@H](COP(O)(O)=O)[C@@H](O)[C@H](O)[C@H](O)CO...   
6494      DB03513              [H][C@](N)(CC1=CNC2=C1[Se]C=C2)C(O)=O   
6495      DB07779  [H][C@](N)(CC(=O)N1CCC[C@@]1([H])CNC(=O)C1=CC=...   
6496      DB03327  [H][C@@](C)(NCC1=C(COP(O)(O)=O)C=NC(C)=C1O)P(O...   
6497      DB04758  CCS(=O)(=O)N[C@H](CC1=CNC2=CC=CC=C12)C(=O)N[C@...   
6498      DB03182  [H]N([H])C1=C2N=CN([C@@H]3O[C@H](COP(O)(=O)OP(...   
6499      DB02816        [H][C@@]1(SC=C(NC=C1C=O)C(O)=O)C1=CN(C)N=N1   
6500      DB07481            CC(C)(C)OC(=O)NCC1=CNC2=C1C(=O)NC(N)=N2   
6501      DB03150    CC1=CN([C@H]2CC[C@@H](COP(O)(O)=O)O2)C(=O)NC1=O   
6502      DB04289  O=C(N1CCN(CC1)C1C2=CC=CC=C2C2=CC=CC=C12)C1=CC2...   
6503      DB08469       CC(C)(C)OC(=O)N1CCC(CC1)C1=NC(=NO1)C1=CC=CS1   
6504      DB07839  [H][C@@](CC1CCCCC1)(NC1=NC2=CC=CC=C2O1)C(=O)NC...   
6505      DB04701      CSC[C@H](NC(=O)CC[C@H](N)C(O)=O)C(=O)NCC(O)=O   
6506      DB07604  [H][C@@]12CC3=C4C(NC=C4[C@]1([H])[C@]1([H])N(C...   
6507      DB01773  CN(CCCC1CCCCC1)C(=O)[C@H](CCC(O)=O)NC(=O)N(CC(...   
6508      DB08419           [H][C@](CC)(COC1=CC=CC=C1)O[P@](C)(Cl)=O   
6509      DB08485  [H][C@@]1(O)C[C@@](O)(C=C(C2=CC=CC(SC3=CC=CC=C...   
6510      DB07262                NC(=N)NC(=N)SCC1=CC=CC(=C1)C(F)(F)F   
6511      DB07030  OC(=O)COC1=CC(Cl)=CC=C1C(=O)NCC1=CC(=CC=C1)[N+...   
6512      DB05364  C1CN2C(=CC=C2C(=O)C3=CC=CC=C3)C1C(=O)O.C(C(CO)...   
6513      DB04746                       CCCCC\C=C/C=C/CCCCCCCCC(O)=O   
6514      DB01703  N1(C(=O)C=CC1=O)CCC12[Fe]3456789%10C%11C3C4C5C...   
6515      DB07521  [H][C@]1(CC\N=C2/NC[C@]([H])(Cl)N(CCNCC3=CSC4=...   
6516      DB03445  [H]\C(N[C@@]([H])(C(O)=O)[C@](C)(CN1C=CN=N1)S(...   
6517      DB11200                O.O.[Al+3].[Cl-].[Zr+4].NCC([O-])=O   
6518      DB01897  O=C1NNC(=O)C2=CC(=CC=C12)[N+]1=NC(\C=C\C2=CC=C...   
6519      DB08404  [H][C@](O)(C(=O)NCCC(=O)NCCSC(=O)CCCCC)C(C)(C)...   
6520      DB05455  CCOC1=CC2=C(C=C1)N(C(=O)C23CCC(CC3)OCCN4CCOCC4...   

      embedding_0  embedding_1  embedding_2  embedding_3  embedding_4  \
6471    -0.472310     0.647877    -0.089998     0.637965     0.650257   
6472     0.526832     0.587521    -0.611328     0.652814     0.554662   
6473     0.638514     0.666598    -0.514444     0.568460     0.529452   
6474     0.583121     0.659789    -0.581

In [23]:
result_df.shape

(6521, 402)

The shape show that there are 6521 rows. There are no missing values, but let's be sure.

In [24]:
# check for missing values
result_df.isna().sum().sum()

0

In [25]:
# save to a CSV file
result_df.to_csv("smiles_embeddings_infer_drugs.csv")